src/datasets.c

```diff
@@ -532,6 +533,62 @@ Dataset *DatasetGet(const char *name, enum DatasetTypes type,
     return NULL;
 }
 
+static bool DatasetIsStatic(Dataset *set)
+{
+    // A set is static if it does not have any dynamic properties like
+    // save and/or state defined.
+    if (set->save == NULL || strcmp(set->save, "") == 0) {
+        return true;
+    }
+    return false;
+}
+
+void DatasetReload(void)
+{
+    // In order to reload the datasets, just mark the current sets as hidden
+    // and clean them up later.
+    // New datasets shall be created with the rule reload and do not require
+    // any intervention.
+    SCMutexLock(&sets_lock);
+    Dataset *set = sets;
+    SCLogInfo("Inside reload function. Sets: %p", sets);
+    while (set != NULL) {
+        if (!DatasetIsStatic(set) || set->from_yaml == true) {
+            SCLogDebug("Not a static set, skipping %s", set->name);
+            set = set->next;
+            continue;
+        }
+        set->hidden = true;
+        SCLogDebug("Set %s at %p hidden successfully", set->name, set);
+        set = set->next;
+    }
+    SCMutexUnlock(&sets_lock);
+}
+
+void DatasetPostReloadCleanup(void)
+{
+    SCLogDebug("Post Reload Cleanup starting.. Hidden sets will be removed");
+    SCMutexLock(&sets_lock);
+    Dataset *cur = sets;
+    Dataset *prev = NULL;
+    while (cur) {
+        Dataset *next = cur->next;
+        if (cur->hidden == false) {
+            prev = cur;
+            cur = next;
+            continue;
+        }
+        // Delete the set in case it was hidden
+        if (prev != NULL) {
```

Comment:
I think this will not work correctly if the list head is removed, as it looks like `sets` is not updated.

Reply:
Yeah don't know what I was looking at while testing it. I put a lock on sets for a reason! Coming up fixed in the next PR